optuna/storages/cached_storage.py

```diff
@@ -24,12 +26,18 @@ def __init__(self) -> None:
         self.system_attrs = dict()  # type: Dict[str, Any]
         self.params = dict()  # type: Dict[str, Any]
         self.distributions = dict()  # type: Dict[str, distributions.BaseDistribution]
+        self.datetime_complete = None  # type: Optional[datetime.datetime]
 
 
 class _StudyInfo:
     def __init__(self) -> None:
-        self.trials = dict()  # type: Dict[int, FrozenTrial]
+        # Trial number to corresponding FrozenTrial.
+        self.trials = {}  # type: Dict[int, FrozenTrial]
+        # A list of trials which do not require storage access to read latest attributes.
+        self.cached_trial_ids = set()  # type: Set[int]
```

Comment:
Do you think it's preferable keeping this bookkeeping `set` even after changing the data structure of `self.trials` from a `list` to a `dict`?

Reply:
The functionality is necessary even after we use `dict` instead of `list`. I think we should change the name of `cached_trial_ids` to something like `owned_or_immutable_trial_ids`, which keeps track of `RUNNING` trials created by this worker or finished trials created by any workers.
`self.trials` caches trials at least as-latest-as the time the `load` method (though it's still not implemented). However, some of the trials might be modified after they are added to the cache. In `get_all_trials` method, the `_CachedStorage` class caches trials into `self.trials` without checking trials' state. Thus, it can contain `RUNNING` trials owned by other workers. For such trials, we need to update the cache via `RDBStorage._get_trials(..., exclude_tiral_ids=...)` method on every `get_all_trials` method call.
(Note: After we implement an incremental-load feature, I think we can remove the necessity to track the `cached_trial_ids`. But it will be after v1.5 and might be after v2.0.)

Reply:
I changed the name of `cached_trial_ids` to `owned_or_finished_trial_ids`.

Reply:
Thank you for clarifying. 🙇